File: go/yagni.go
// YAGNI (You Ain't Gonna Need It)
// Don't build a cannon to kill a fly; you might not even need it afterward.

package main

import "fmt"

// Before
type UserBefore struct {
	Name        string
	Email       string
	Age         *int    // Unused feature
	Address     *string // Unused feature
	PhoneNumber *string // Unused feature
}

func NewUserBefore(name, email string) *UserBefore {
	return &UserBefore{
		Name:        name,
		Email:       email,
		Age:         nil,
		Address:     nil,
		PhoneNumber: nil,
	}
}

func (u *UserBefore) SendWelcomeEmail() {
	fmt.Printf("Welcome, %s! A welcome email has been sent to %s.\n", u.Name, u.Email)
}

// After
type User struct {
	Name  string
	Email string
}

func NewUser(name, email string) *User {
	return &User{
		Name:  name,
		Email: email,
	}
}

func (u *User) SendWelcomeEmail() {
	fmt.Printf("Welcome, %s! A welcome email has been sent to %s.\n", u.Name, u.Email)
}

func main() {
	userBefore := NewUserBefore("Alice", "alice@example.com")
	userBefore.SendWelcomeEmail()

	fmt.Println()

	user := NewUser("Alice", "alice@example.com")
	user.SendWelcomeEmail()
}

File: kotlin/yagni.kt
// YAGNI (You Ain't Gonna Need It)
// Don't build a cannon to kill a fly; you might not even need it afterward.

// Before
class UserBefore(
    val name: String,
    val email: String
) {
    var age: Int? = null  // Unused feature
    var address: String? = null  // Unused feature
    var phoneNumber: String? = null  // Unused feature

    fun sendWelcomeEmail() {
        println("Welcome, $name! A welcome email has been sent to $email.")
    }
}

// After
class User(
    val name: String,
    val email: String
) {
    fun sendWelcomeEmail() {
        println("Welcome, $name! A welcome email has been sent to $email.")
    }
}

fun main() {
    val userBefore = UserBefore("Alice", "alice@example.com")
    userBefore.sendWelcomeEmail()

    println()

    val user = User("Alice", "alice@example.com")
    user.sendWelcomeEmail()
}
